src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1282,49 +1282,62 @@ private async ValueTask<(Stream?, TransportContext?, HttpResponseMessage?)> Conn
+        private async ValueTask<Stream> ConnectToTcpHostAsync(string host, int port, HttpRequestMessage initialRequest, bool async, CancellationToken cancellationToken)
         {
-            Socket socket = new Socket(SocketType.Stream, ProtocolType.Tcp);
-            socket.NoDelay = true;
+            cancellationToken.ThrowIfCancellationRequested();
 
+            var endPoint = new DnsEndPoint(host, port);
+            Socket? socket = null;
             try
             {
-                await socket.ConnectAsync(context.DnsEndPoint, cancellationToken).ConfigureAwait(false);
-                return new NetworkStream(socket, ownsSocket: true);
-            }
-            catch
-            {
-                socket.Dispose();
-                throw;
-            }
-        }
-
-        private static readonly Func<SocketsHttpConnectionContext, CancellationToken, ValueTask<Stream>> s_defaultConnectCallback = DefaultConnectAsync;
-
-        private ValueTask<Stream> ConnectToTcpHostAsync(string host, int port, HttpRequestMessage initialRequest, bool async, CancellationToken cancellationToken)
-        {
-            if (async)
-            {
-                Func<SocketsHttpConnectionContext, CancellationToken, ValueTask<Stream>> connectCallback = Settings._connectCallback ?? s_defaultConnectCallback;
+                // If a ConnectCallback was supplied, use that to establish the connection.
+                if (Settings._connectCallback != null)
+                {
+                    ValueTask<Stream> streamTask = Settings._connectCallback(new SocketsHttpConnectionContext(endPoint, initialRequest), cancellationToken);
 
-                var endPoint = new DnsEndPoint(host, port);
-                return ConnectHelper.ConnectAsync(connectCallback, endPoint, initialRequest, cancellationToken);
-            }
+                    Stream stream;
+                    if (async || streamTask.IsCompleted)
+                    {
+                        stream = await streamTask.ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        // User-provided ConnectCallback is completing asynchronously but the user is making a synchronous request; if the user cares, they should
+                        // set it up so that synchronous requests are made on a handler with a synchronously-completing ConnectCallback supplied. If in the future,
+                        // we could add a Boolean to SocketsHttpConnectionContext (https://github.com/dotnet/runtime/issues/44876) to let the callback know whether
+                        // this request is sync or async.  For now, log it and block.
+                        Trace($"{nameof(SocketsHttpHandler.ConnectCallback)} completing asynchronously for a synchronous request.");
+                        stream = streamTask.AsTask().GetAwaiter().GetResult();
```

Comment:
Shouldn't we rather throw instead of blocking for sync?

Reply:
That was what I did initially, but I think that's problematic because it's non-deterministic: if the operation happens to complete by the time we check, it won't throw, but if it doesn't, it will.  That then makes it much more likely you'll only experience the failures once you deploy.  On top of that (and I've saved dealing with this for a separate follow-up PR), we shipped .NET 5 without such failures for the plaintext stream callback, so today with .NET 5, you can hook up a callback that asynchronously completes, and it'll work with sync methods; in a debug build of System.Net.Http it'll trigger an assert, but in release builds it'll just end up blocking until completion.  I see these two callbacks as needing to behave similarly, and so I think the right answer is to just log both and let them block.  We already block in other circumstances as well, e.g. if you have a MaxConnectionsPerServer set, we'll block waiting for an available connection for sync operations, rather than throwing, so not throwing here is consistent with that, too.

Reply:
I guess my original concern will get addressed with the proper implementation of https://github.com/dotnet/runtime/issues/44876, and you're reasoning makes sense here. So thanks.